src/inventory.rs

```diff
@@ -0,0 +1,197 @@
+    /// hotbar, excluding offhand. These slots are shown when the player is
+    /// looking at another inventory.
+    pub fn general_slots() -> Range<SlotId> {
+        9..45
+    }
+
+    pub fn hotbar_slots() -> Range<SlotId> {
+        36..45
+    }
+
+    pub fn hotbar_to_slot(hotbar_slot: i16) -> Option<SlotId> {
+        if !(0..=8).contains(&hotbar_slot) {
+            return None;
+        }
+
+        Some(Self::hotbar_slots().start + hotbar_slot)
+    }
+}
+
+impl Default for PlayerInventory {
+    fn default() -> Self {
+        Self {
+            // Can't do the shorthand because Slot is not Copy.
+            slots: [
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+            ],
+        }
+    }
+}
+
+impl Inventory for PlayerInventory {
+    fn get_slot(&self, slot_id: SlotId) -> &Slot {
+        if slot_id < 0 {
+            // TODO: dont panic
+            panic!("invalid slot id")
+        }
+        &self.slots[slot_id as usize]
+    }
+
+    fn set_slot(&mut self, slot_id: SlotId, slot: Slot) {
+        if slot_id < 0 {
+            // TODO: dont panic
+            panic!("invalid slot id")
+        }
+        self.slots[slot_id as usize] = slot;
+    }
+
+    fn capacity(&self) -> usize {
+        self.slots.len()
+    }
+}
+
+/// Represents what the player sees when they open an object's Inventory.
+///
+/// This exists because when an object inventory screen is being shown to the
+/// player, it also shows part of the player's inventory so they can move items
+/// between the inventories.
+#[derive(Debug)]
+pub struct WindowInventory<T>
+where
+    T: Inventory,
```

Comment:
Nitpicking but I don't think it's idiomatic to put a trait bound on a struct definition unless it _needs_ to be that trait for the definition to work. Also maybe you can call the generic type `I` instead of `T` in this module.

Reply:
The trait bound is necessary. `WindowInventory` acts like a translation layer between the 2 displayed inventories. In order to do that, the other object needs to impl `Inventory`.

Reply:
I mean the bound in the struct definition. Not in the later `impl`.